x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.utils.persistence;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.bulk.BulkItemResponse;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.support.WriteRequest;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Randomness;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Random;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+
+public class ResultsPersisterService {
+    private static final Logger LOGGER = LogManager.getLogger(ResultsPersisterService.class);
+
+    public static final Setting<Integer> PERSIST_RESULTS_MAX_RETRIES = Setting.intSetting(
+        "xpack.ml.persist_results_max_retries",
+        15,
+        0,
+        50,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope);
+    private static final int MAX_RETRY_SLEEP_MILLIS = (int)Duration.ofMinutes(15).toMillis();
+    private static final int MIN_RETRY_SLEEP_MILLIS = 50;
+    // Having an exponent higher than this causes integer overflow
+    private static final int MAX_RETRY_EXPONENT = 29;
```

Comment:
I think this will need changing to 24, otherwise the `int` will overflow when multiplied by 50

Reply:
```
jshell> int val = ((1 << 29) - 1) * (100 / 2);
val ==> 1073741774
```

Reply:
As a `long` the answer is 26843545550 though, so 1073741774 is due to wrapping.

Try `int val = ((1 << 27) - 1) * (100 / 2);`

If we're going to rely on wrapping then it would probably be clearer to just say `if (currentAttempt > SOMETHING) { max = magic number }`

Reply:
You are 100% right. I reduced to 24.